rust/crd/src/lib.rs

```diff
@@ -73,6 +91,55 @@ impl KafkaVersion {
     pub fn fully_qualified_version(&self) -> String {
         format!("{}-{}", self.scala_version(), self.kafka_version())
     }
+
+    pub fn parse(&self) -> Result<(String, Version), error::Error> {
+        // TODO: careful with the string comparison. Works for 2.12 and 2.13.
+        //   will fail for 2.2 and 2.13. At this point we need a smarter comparison.
+        if !Self::SUPPORTED_SCALA_VERSIONS.contains(&self.scala_version()) {
+            return Err(Error::ScalaVersionNotSupported {
+                scala_version: self.scala_version().to_string(),
+                full_version: self.fully_qualified_version(),
+                supported_versions: &Self::SUPPORTED_SCALA_VERSIONS,
+            });
+        }
+
+        match Version::parse(self.kafka_version()) {
+            Ok(kafka_version) => Ok((self.scala_version().to_string(), kafka_version)),
+            Err(err) => Err(Error::KafkaVersionNotParseable {
+                version: self.fully_qualified_version(),
+                error: err.to_string(),
+            }),
+        }
+    }
+}
+
+impl Versioning for KafkaVersion {
+    fn versioning_state(&self, other: &Self) -> VersioningState {
+        let (from_scala, from_kafka) = match self.parse() {
+            Ok((scala, kafka)) => (scala, kafka),
+            Err(err) => return VersioningState::Invalid(err.to_string()),
+        };
+
+        let (to_scala, to_kafka) = match other.parse() {
+            Ok((scala, kafka)) => (scala, kafka),
+            Err(err) => return VersioningState::Invalid(err.to_string()),
+        };
+
+        match (from_scala.cmp(&to_scala), from_kafka.cmp(&to_kafka)) {
+            // scala version, kafka_version
+            (Ordering::Equal, Ordering::Equal) => VersioningState::NoOp,
+
+            (Ordering::Less, Ordering::Less)
+            | (Ordering::Less, Ordering::Equal)
+            | (Ordering::Equal, Ordering::Less)
+            | (Ordering::Greater, Ordering::Less) => VersioningState::ValidUpgrade,
```

Comment:
IMO the scala version is relevant for the underlying JRE/JDK that but not for Kafka upgrades or downgrades.

Therefore I suggest we just look at the Kafka version and have make plan regarding Java:
```suggestion
            _, Ordering::Less => VersioningState::ValidUpgrade,
```